### DISCUSSION TITLE
Code isolation

### DISCUSSION THREAD

**Discussion by hexwit** [lit/lit#3101] [Q&A]
Hello!

I have a goal: show modal window (with some custom logic) on any page via browser extension. I click extension button - modal window with business logic injected into the current page and displayed its content.

Here I have a potential problem: how to isolate JS logic in my component to avoid any interfering with other code on the page?
Is there any technique how I can use, for example, lodash in my injected component, and do not conflict with lodash that was available on the page, but had another version. So I need isolation for my own code. 

Is there any technique to protect my code, and code on the page from interfering?

Thanks.

**Comment by justinfagnani** [comment]
Is your goal preventing code from accidentally interfering with other code, or security?

If you're going for security - as in: you're running untrusted code and need to protect the page from it - you'll need to at least use an iframe from a different origin.

Otherwise, for most code, if you're using JS modules, you should be fine. Two copies of a module will run independently unless it goes out of its way to mess with global state. Even non-standard module formats like CommonJS or UMD behave well if bundled separately.

Custom elements currently pose another problem, because they use a shared global resource with the custom elements registry. For that you probably want to use the custom elements polyfill inside the content script.
